db/memtable_list.cc

```diff
@@ -269,10 +269,243 @@ void MemTableListVersion::TrimHistory(autovector<MemTable*>* to_delete) {
+    //                    +-----------------+ |   +-----------------+
+    //                            |           |
+    //                            V           |
+    //                    +-----------------+ |
+    // imm_lists[n3]:     |mems_list[M-1][0]| |
+    // ...                +-----------------+ |
+    //                            |           |
+    //                            +-----------+
+    // imm_lists[N - 1]:
+    //
+    MemTable* mem = nullptr;
+    MemTable* head = nullptr;
+    for (const auto& mems : mems_list) {
+      if (!mems->empty()) {
+        if (head == nullptr) {
+          head = mems->front();
+        }
+        if (mem != nullptr) {
+          mem->next_ = mems->front();
+        }
+        mem = mems->front();
+      }
+    }
+    if (mem != nullptr) {
+      mem->next_ = head;
+    }
+  }
+
+  int num = static_cast<int>(imm_lists.size());
+  assert(num == static_cast<int>(cfds.size()));
+  for (int k = 0; k != num; ++k) {
+    for (int i = 0; i != static_cast<int>(mems_list[k]->size()); ++i) {
+      assert(i == 0 || (*mems_list[k])[i]->GetEdits()->NumEntries() == 0);
+      (*mems_list[k])[i]->flush_completed_ = true;
+      (*mems_list[k])[i]->file_number_ = file_meta[k].fd.GetNumber();
+    }
+  }
+
+  Status s;
+  if (atomic_flush_commit_in_progress_) {
+    return s;
+  }
+
+  atomic_flush_commit_in_progress_ = true;
+
+  while (s.ok()) {
+    std::unordered_map<MemTable*, int> mem_to_idx;
+    autovector<int> next;
+    for (int i = 0; i != num; ++i) {
+      const std::list<MemTable*>& memlist = imm_lists[i]->current_->memlist_;
+      if (!memlist.empty()) {
+        const auto it = memlist.rbegin();
+        if ((*it)->flush_completed_) {
+          mem_to_idx.insert({*it, i});
+        }
+      }
+      next.push_back(-1);
+    }
+    if (mem_to_idx.empty()) {
+      // Nothing to commit
+      break;
+    }
+    for (int i = 0; i != num; ++i) {
+      const std::list<MemTable*>& memlist = imm_lists[i]->current_->memlist_;
+      if (!memlist.empty()) {
+        const auto it = memlist.rbegin();
+        assert((*it)->next_ != nullptr);
+        if ((*it)->flush_completed_) {
+          const auto next_it = mem_to_idx.find((*it)->next_);
+          if (next_it != mem_to_idx.end()) {
+            next[i] = next_it->second;
+          }
+        }
+      }
+    }
+    while (true) {
+      // Try to find a linked list formed by memtables at the rbegin side of
+      // each imm_lists. If such a linked list exist, we have found a new group
+      // to commit to MANIFEST.
+      int k = 0;
+      while (k < num && next[k] < 0) {
+        ++k;
+      }
+      if (k == num) {
+        break;
+      }
+
+      autovector<int> idx;
+      int head = k;
+      do {
+        idx.push_back(k);
+        k = next[k];
+      } while (k != head && k >= 0);
+
+      for (auto i : idx) {
+        next[i] = -1;
+      }
+
+      if (k < 0) {
```

Comment:
When does this happen?

Reply:
`next[j] < 0` can happen in two cases:
* imm_lists[i] has already been committed in this iteration, or
* Some memtables in the same atomic flush are not ready to commit because they are not the oldest in their corresponding imm_list.

When `k < 0` happens, it means the current `next` array does not contain the information of any ready atomic flush, thus needs to be updated. That's why we start a new iteration to populate `next`.